File: sdk/rust/src/filter.rs
//! Fluent builder for RSQL filter expressions, accepted by the `filter`
//! parameter of the catalog and scrobble-feed queries
//! (`app.rocksky.song.getSongs`, `app.rocksky.artist.getArtists`,
//! `app.rocksky.album.getAlbums`, `app.rocksky.scrobble.getScrobbles`).
//!
//! ```
//! use rocksky::Filter;
//!
//! let filter = Filter::eq("artist", "Daft Punk")
//!     .and(Filter::gt("duration", 200_000))
//!     .or(Filter::is_in("genre", ["house", "electro"]));
//!
//! assert_eq!(
//!     filter.build(),
//!     "artist==\"Daft Punk\";duration=gt=200000,genre=in=(house,electro)",
//! );
//! ```
//!
//! String values are quoted and escaped automatically when they contain
//! characters RSQL reserves; `*` wildcards pass through unquoted so
//! `Filter::eq("artist", "Daft*")` performs a case-insensitive match.

use std::fmt;

/// A single value in an RSQL comparison. Constructed implicitly via
/// `Into<FilterValue>` from `&str`, `String`, `i64`, `i32`, `u32`, `f64`
/// and `bool`.
#[derive(Debug, Clone, PartialEq)]
pub enum FilterValue {
    /// A string value; quoted/escaped on render when needed.
    Str(String),
    /// An integer value; rendered bare, without decimals.
    Int(i64),
    /// A floating-point value; rendered with Rust's minimal `Display` repr.
    Float(f64),
    /// A boolean; rendered as `true` / `false`.
    Bool(bool),
}

impl From<&str> for FilterValue {
    fn from(v: &str) -> Self {
        FilterValue::Str(v.to_owned())
    }
}
impl From<String> for FilterValue {
    fn from(v: String) -> Self {
        FilterValue::Str(v)
    }
}
impl From<i64> for FilterValue {
    fn from(v: i64) -> Self {
        FilterValue::Int(v)
    }
}
impl From<i32> for FilterValue {
    fn from(v: i32) -> Self {
        FilterValue::Int(v as i64)
    }
}
impl From<u32> for FilterValue {
    fn from(v: u32) -> Self {
        FilterValue::Int(v as i64)
    }
}
impl From<f64> for FilterValue {
    fn from(v: f64) -> Self {
        FilterValue::Float(v)
    }
}
impl From<bool> for FilterValue {
    fn from(v: bool) -> Self {
        FilterValue::Bool(v)
    }
}

/// Characters that never need quoting in an RSQL value
/// (`*` kept bare so wildcards work).
fn is_safe_char(c: char) -> bool {
    c.is_ascii_alphanumeric() || matches!(c, '_' | '.' | ':' | '@' | '*' | '+' | '-')
}

fn render_value(value: &FilterValue) -> String {
    match value {
        FilterValue::Int(n) => n.to_string(),
        FilterValue::Float(n) => n.to_string(),
        FilterValue::Bool(b) => b.to_string(),
        FilterValue::Str(s) => {
            if !s.is_empty() && s.chars().all(is_safe_char) {
                s.clone()
            } else {
                let escaped = s.replace('\\', "\\\\").replace('"', "\\\"");
                format!("\"{escaped}\"")
            }
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum NodeKind {
    Comparison,
    And,
    Or,
}

/// Fluent builder for RSQL filter expressions. See the [module docs](self).
#[derive(Debug, Clone)]
pub struct Filter {
    expr: String,
    kind: NodeKind,
}

impl Filter {
    fn comparison(field: &str, op: &str, value: FilterValue) -> Self {
        Filter {
            expr: format!("{field}{op}{}", render_value(&value)),
            kind: NodeKind::Comparison,
        }
    }

    fn list<I, V>(field: &str, op: &str, name: &str, values: I) -> Self
    where
        I: IntoIterator<Item = V>,
        V: Into<FilterValue>,
    {
        let rendered: Vec<String> = values
            .into_iter()
            .map(|v| render_value(&v.into()))
            .collect();
        if rendered.is_empty() {
            panic!("Filter::{name}(\"{field}\", ...) needs at least one value");
        }
        Filter {
            expr: format!("{field}{op}({})", rendered.join(",")),
            kind: NodeKind::Comparison,
        }
    }

    /// `field==value` — equals; `*` in string values is a wildcard.
    pub fn eq(field: impl AsRef<str>, value: impl Into<FilterValue>) -> Self {
        Filter::comparison(field.as_ref(), "==", value.into())
    }

    /// `field!=value` — not equals.
    pub fn ne(field: impl AsRef<str>, value: impl Into<FilterValue>) -> Self {
        Filter::comparison(field.as_ref(), "!=", value.into())
    }

    /// `field=gt=value` — greater than.
    pub fn gt(field: impl AsRef<str>, value: impl Into<FilterValue>) -> Self {
        Filter::comparison(field.as_ref(), "=gt=", value.into())
    }

    /// `field=ge=value` — greater than or equal.
    pub fn ge(field: impl AsRef<str>, value: impl Into<FilterValue>) -> Self {
        Filter::comparison(field.as_ref(), "=ge=", value.into())
    }

    /// `field=lt=value` — less than.
    pub fn lt(field: impl AsRef<str>, value: impl Into<FilterValue>) -> Self {
        Filter::comparison(field.as_ref(), "=lt=", value.into())
    }

    /// `field=le=value` — less than or equal.
    pub fn le(field: impl AsRef<str>, value: impl Into<FilterValue>) -> Self {
        Filter::comparison(field.as_ref(), "=le=", value.into())
    }

    /// `field=in=(a,b)` — matches any of the values.
    ///
    /// # Panics
    /// Panics if `values` is empty.
    pub fn is_in<I, V>(field: impl AsRef<str>, values: I) -> Self
    where
        I: IntoIterator<Item = V>,
        V: Into<FilterValue>,
    {
        Filter::list(field.as_ref(), "=in=", "is_in", values)
    }

    /// `field=out=(a,b)` — matches none of the values.
    ///
    /// # Panics
    /// Panics if `values` is empty.
    pub fn is_out<I, V>(field: impl AsRef<str>, values: I) -> Self
    where
        I: IntoIterator<Item = V>,
        V: Into<FilterValue>,
    {
        Filter::list(field.as_ref(), "=out=", "is_out", values)
    }

    /// `field==null` — the field is NULL.
    pub fn is_null(field: impl AsRef<str>) -> Self {
        Filter {
            expr: format!("{}==null", field.as_ref()),
            kind: NodeKind::Comparison,
        }
    }

    /// `field!=null` — the field is not NULL.
    pub fn is_not_null(field: impl AsRef<str>) -> Self {
        Filter {
            expr: format!("{}!=null", field.as_ref()),
            kind: NodeKind::Comparison,
        }
    }

    /// Both sides must match (`;`). An `or` operand is parenthesized to keep
    /// RSQL precedence.
    pub fn and(self, other: Filter) -> Self {
        Filter {
            expr: format!("{};{}", self.render_in_and(), other.render_in_and()),
            kind: NodeKind::And,
        }
    }

    /// Either side may match (`,`).
    pub fn or(self, other: Filter) -> Self {
        Filter {
            expr: format!("{},{}", self.expr, other.expr),
            kind: NodeKind::Or,
        }
    }

    fn render_in_and(&self) -> String {
        if self.kind == NodeKind::Or {
            format!("({})", self.expr)
        } else {
            self.expr.clone()
        }
    }

    /// The RSQL expression string to send as the `filter` query param.
    pub fn build(self) -> String {
        self.expr
    }
}

impl fmt::Display for Filter {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.expr)
    }
}

impl From<Filter> for String {
    fn from(filter: Filter) -> Self {
        filter.expr
    }
}

#[cfg(test)]
mod tests {
    use super::Filter;

    #[test]
    fn eq_bare() {
        assert_eq!(Filter::eq("artist", "Radiohead").build(), "artist==Radiohead");
    }

    #[test]
    fn eq_quoted_space() {
        assert_eq!(
            Filter::eq("artist", "Daft Punk").build(),
            "artist==\"Daft Punk\""
        );
    }

    #[test]
    fn eq_escapes_embedded_quotes() {
        assert_eq!(
            Filter::eq("title", r#"He said "hi""#).build(),
            "title==\"He said \\\"hi\\\"\""
        );
    }

    #[test]
    fn eq_wildcard_unquoted() {
        assert_eq!(Filter::eq("artist", "Daft*").build(), "artist==Daft*");
    }

    #[test]
    fn ne() {
        assert_eq!(Filter::ne("artist", "Eminem").build(), "artist!=Eminem");
    }

    #[test]
    fn gt() {
        assert_eq!(Filter::gt("duration", 200000).build(), "duration=gt=200000");
    }

    #[test]
    fn ge() {
        assert_eq!(Filter::ge("year", 2000).build(), "year=ge=2000");
    }

    #[test]
    fn lt() {
        assert_eq!(Filter::lt("trackNumber", 5).build(), "trackNumber=lt=5");
    }

    #[test]
    fn le() {
        assert_eq!(Filter::le("year", 1999).build(), "year=le=1999");
    }

    #[test]
    fn is_in() {
        assert_eq!(
            Filter::is_in("genre", ["house", "electro"]).build(),
            "genre=in=(house,electro)"
        );
    }

    #[test]
    fn is_out_quotes_unsafe_values() {
        assert_eq!(
            Filter::is_out("genre", ["hip hop"]).build(),
            "genre=out=(\"hip hop\")"
        );
    }

    #[test]
    #[should_panic(expected = "needs at least one value")]
    fn is_in_empty_panics() {
        let _ = Filter::is_in("genre", Vec::<String>::new());
    }

    #[test]
    #[should_panic(expected = "needs at least one value")]
    fn is_out_empty_panics() {
        let _ = Filter::is_out("genre", Vec::<String>::new());
    }

    #[test]
    fn is_null() {
        assert_eq!(Filter::is_null("uri").build(), "uri==null");
    }

    #[test]
    fn is_not_null() {
        assert_eq!(Filter::is_not_null("uri").build(), "uri!=null");
    }

    #[test]
    fn and_chain() {
        assert_eq!(
            Filter::eq("artist", "Radiohead")
                .and(Filter::gt("duration", 200000))
                .build(),
            "artist==Radiohead;duration=gt=200000"
        );
    }

    #[test]
    fn or_chain() {
        assert_eq!(
            Filter::eq("artist", "Radiohead")
                .or(Filter::eq("artist", "Muse"))
                .build(),
            "artist==Radiohead,artist==Muse"
        );
    }

    #[test]
    fn or_then_and_parenthesizes_left_or() {
        assert_eq!(
            Filter::eq("artist", "Radiohead")
                .or(Filter::eq("artist", "Muse"))
                .and(Filter::gt("duration", 200000))
                .build(),
            "(artist==Radiohead,artist==Muse);duration=gt=200000"
        );
    }

    #[test]
    fn and_with_or_operand_parenthesizes_right_or() {
        assert_eq!(
            Filter::eq("artist", "Radiohead")
                .and(Filter::eq("genre", "house").or(Filter::eq("genre", "electro")))
                .build(),
            "artist==Radiohead;(genre==house,genre==electro)"
        );
    }

    #[test]
    fn and_then_or_adds_no_parentheses() {
        assert_eq!(
            Filter::eq("artist", "Radiohead")
                .and(Filter::gt("duration", 200000))
                .or(Filter::eq("genre", "house"))
                .build(),
            "artist==Radiohead;duration=gt=200000,genre==house"
        );
    }

    #[test]
    fn dotted_field() {
        assert_eq!(
            Filter::eq("track.artist", "Daft Punk").build(),
            "track.artist==\"Daft Punk\""
        );
    }

    #[test]
    fn bool_value() {
        assert_eq!(Filter::eq("liked", true).build(), "liked==true");
    }

    #[test]
    fn display_and_from_string() {
        let f = Filter::eq("artist", "Radiohead");
        assert_eq!(f.to_string(), "artist==Radiohead");
        let s: String = f.into();
        assert_eq!(s, "artist==Radiohead");
    }
}

File: sdk/typescript/src/filter.ts
/**
 * Fluent builder for RSQL filter expressions, accepted by the `filter`
 * parameter of the catalog and scrobble-feed queries
 * (app.rocksky.song.getSongs, app.rocksky.artist.getArtists,
 * app.rocksky.album.getAlbums, app.rocksky.scrobble.getScrobbles).
 *
 * ```ts
 * import { Filter } from "@rocksky/sdk";
 *
 * const filter = Filter.eq("artist", "Daft Punk")
 *   .and(Filter.gt("duration", 200_000))
 *   .or(Filter.in("genre", ["house", "electro"]));
 *
 * await client.catalogSongs(50, 0, undefined, filter);
 * // artist=="Daft Punk";duration=gt=200000,genre=in=(house,electro)
 * ```
 *
 * String values are quoted and escaped automatically when they contain
 * characters RSQL reserves; `*` wildcards pass through unquoted so
 * `Filter.eq("artist", "Daft*")` performs a case-insensitive match.
 */
export type FilterValue = string | number | boolean;

/** Characters that never need quoting in an RSQL value (`*` kept bare so wildcards work). */
const SAFE_VALUE = /^[A-Za-z0-9_.:@*+-]+$/;

const renderValue = (value: FilterValue): string => {
  if (typeof value === "number" || typeof value === "boolean") return String(value);
  if (value.length > 0 && SAFE_VALUE.test(value)) return value;
  return `"${value.replace(/\\/g, "\\\\").replace(/"/g, '\\"')}"`;
};

type NodeKind = "comparison" | "and" | "or";

export class Filter {
  private constructor(
    private readonly expr: string,
    private readonly kind: NodeKind,
  ) {}

  private static comparison(field: string, op: string, value: FilterValue): Filter {
    return new Filter(`${field}${op}${renderValue(value)}`, "comparison");
  }

  private static list(field: string, op: string, values: FilterValue[]): Filter {
    if (values.length === 0) {
      throw new Error(`Filter.${op === "=in=" ? "in" : "out"}("${field}", ...) needs at least one value`);
    }
    return new Filter(`${field}${op}(${values.map(renderValue).join(",")})`, "comparison");
  }

  /** `field==value` — equals; `*` in string values is a wildcard. */
  static eq(field: string, value: FilterValue): Filter {
    return Filter.comparison(field, "==", value);
  }

  /** `field!=value` — not equals. */
  static ne(field: string, value: FilterValue): Filter {
    return Filter.comparison(field, "!=", value);
  }

  /** `field=gt=value` — greater than. */
  static gt(field: string, value: FilterValue): Filter {
    return Filter.comparison(field, "=gt=", value);
  }

  /** `field=ge=value` — greater than or equal. */
  static ge(field: string, value: FilterValue): Filter {
    return Filter.comparison(field, "=ge=", value);
  }

  /** `field=lt=value` — less than. */
  static lt(field: string, value: FilterValue): Filter {
    return Filter.comparison(field, "=lt=", value);
  }

  /** `field=le=value` — less than or equal. */
  static le(field: string, value: FilterValue): Filter {
    return Filter.comparison(field, "=le=", value);
  }

  /** `field=in=(a,b)` — matches any of the values. */
  static in(field: string, values: FilterValue[]): Filter {
    return Filter.list(field, "=in=", values);
  }

  /** `field=out=(a,b)` — matches none of the values. */
  static out(field: string, values: FilterValue[]): Filter {
    return Filter.list(field, "=out=", values);
  }

  /** `field==null` — the field is NULL. */
  static isNull(field: string): Filter {
    return new Filter(`${field}==null`, "comparison");
  }

  /** `field!=null` — the field is not NULL. */
  static isNotNull(field: string): Filter {
    return new Filter(`${field}!=null`, "comparison");
  }

  /** Both sides must match (`;`). An `or` operand is parenthesized to keep RSQL precedence. */
  and(other: Filter): Filter {
    return new Filter(`${this.renderIn("and")};${other.renderIn("and")}`, "and");
  }

  /** Either side may match (`,`). */
  or(other: Filter): Filter {
    return new Filter(`${this.expr},${other.expr}`, "or");
  }

  private renderIn(parent: "and"): string {
    return this.kind === "or" ? `(${this.expr})` : this.expr;
  }

  /** The RSQL expression string to send as the `filter` query param. */
  build(): string {
    return this.expr;
  }

  toString(): string {
    return this.expr;
  }
}

/** Filterable fields of app.rocksky.song.getSongs. */
export const SongFields = {
  title: "title",
  artist: "artist",
  album: "album",
  albumArtist: "albumArtist",
  genre: "genre",
  composer: "composer",
  label: "label",
  duration: "duration",
  trackNumber: "trackNumber",
  discNumber: "discNumber",
  mbId: "mbId",
  isrc: "isrc",
  sha256: "sha256",
  uri: "uri",
  albumUri: "albumUri",
  artistUri: "artistUri",
  createdAt: "createdAt",
} as const;

/** Filterable fields of app.rocksky.album.getAlbums. */
export const AlbumFields = {
  title: "title",
  artist: "artist",
  year: "year",
  releaseDate: "releaseDate",
  sha256: "sha256",
  uri: "uri",
  artistUri: "artistUri",
  createdAt: "createdAt",
} as const;

/** Filterable fields of app.rocksky.artist.getArtists. */
export const ArtistFields = {
  name: "name",
  genres: "genres",
  bornIn: "bornIn",
  born: "born",
  died: "died",
  sha256: "sha256",
  uri: "uri",
  createdAt: "createdAt",
} as const;

/** Filterable fields of app.rocksky.playlist.getPlaylists (`track.*` selectors match the playlist's contents). */
export const PlaylistFields = {
  name: "name",
  title: "title",
  description: "description",
  uri: "uri",
  spotifyLink: "spotifyLink",
  tidalLink: "tidalLink",
  appleMusicLink: "appleMusicLink",
  createdAt: "createdAt",
  updatedAt: "updatedAt",
  curatorDid: "curatorDid",
  curatorHandle: "curatorHandle",
  curatorName: "curatorName",
  trackTitle: "track.title",
  trackArtist: "track.artist",
  trackAlbum: "track.album",
  trackAlbumArtist: "track.albumArtist",
} as const;

/** Filterable fields of app.rocksky.scrobble.getScrobbles (dotted selectors reach the joined track/user/artist). */
export const ScrobbleFields = {
  uri: "uri",
  date: "date",
  timestamp: "timestamp",
  title: "title",
  artist: "artist",
  album: "album",
  trackTitle: "track.title",
  trackArtist: "track.artist",
  trackAlbum: "track.album",
  trackAlbumArtist: "track.albumArtist",
  trackGenre: "track.genre",
  trackDuration: "track.duration",
  trackIsrc: "track.isrc",
  trackMbId: "track.mbId",
  userDid: "user.did",
  userHandle: "user.handle",
  userDisplayName: "user.displayName",
  artistName: "artist.name",
  artistGenres: "artist.genres",
} as const;
